src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -645,26 +645,29 @@ private Uri GetContainingEntitySetUri(Uri baseUri, ODataUri odataUri)
         {
             ODataPath path = odataUri.Path;
             List<ODataPathSegment> segments = path.ToList();
+            int segmentsCount = segments.Count - 1;
             ODataPathSegment lastSegment = segments.Last();
             while (!(lastSegment is NavigationPropertySegment) && !(lastSegment is OperationSegment))
             {
-                segments.Remove(lastSegment);
-                lastSegment = segments.Last();
+                segmentsCount--;
+                lastSegment = segments[segmentsCount];
             }
 
             // also removed the last navigation property segment
-            segments.Remove(lastSegment);
+            segments = segments.GetRange(0, segmentsCount);
 
             // Remove the unnecessary type cast segment.
             ODataPathSegment nextToLastSegment = segments.Last();
+            int lastIndex = segments.Count - 1;
             while (nextToLastSegment is TypeSegment)
             {
                 ODataPathSegment previousSegment = segments[segments.Count - 2];
                 IEdmStructuredType owningType = previousSegment.TargetEdmType as IEdmStructuredType;
                 if (owningType != null && owningType.FindProperty(lastSegment.Identifier) != null)
                 {
-                    segments.Remove(nextToLastSegment);
+                    segments.RemoveAt(lastIndex);
                     nextToLastSegment = segments.Last();

```

Comment:
>segments.Last() [](start = 40, length = 15)

is it more efficient to do segments[lastIndex]?

Reply:
Let me evaluate this, my assumption is that that segments.Last() and segments[lastindex] should give use the same performance esp on lists.

Reply:
According to [this](https://stackoverflow.com/questions/1377864/what-is-the-performance-of-the-last-extension-method-for-listt), there is an optimization in .Last() that checks to see if it's a list, but that incurs a cast. Not a huge deal, but if you already have the lastIndex, seems it would be faster just to use it.